wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/util/RecipeDownloader.java

```diff
@@ -65,9 +65,12 @@ public RecipeImpl getRecipe(MachineConfig machineConfig) throws MachineException
         try {
             UriBuilder targetUriBuilder = UriBuilder.fromUri(location);
             // add user token to be able to download user's private recipe
-            final String apiEndPointHost = apiEndpoint.getHost();
-            final String host = targetUriBuilder.build().getHost();
-            if (apiEndPointHost.equals(host)) {
+            final URI recipeUri = targetUriBuilder.build();
+            if (!recipeUri.isAbsolute() && recipeUri.getHost() == null) {
+                targetUriBuilder.scheme(apiEndpoint.getScheme())
+                                .host(apiEndpoint.getHost())
+                                .port(apiEndpoint.getPort())
+                                .replacePath(apiEndpoint.getPath() + location);
```

Comment:
Is it possible that api endpoint doesn't have trailing slash? In that case I suppose it would be better to use .replacePath().path()